File: projecteuler/python/p002.py
# https://projecteuler.net/problem=2

from euler.mathtools import fibonacci


def main():
    maxima = 4 * 10 ** 6

    fibs = [0]
    i = 1
    while fibs[-1] < maxima:
        fibs.append(fibonacci(i))
        i += 1

    fibs = fibs[:-1]

    sigma = sum([f for f in fibs if f % 2 == 0])
    print(sigma)


if __name__ == '__main__':
    main()

File: projecteuler/lua/p002.lua
function fib(n)
  local cache = {}
  
  local function _fib(m)
    if m < 2 then
      return m
    end

    if cache[m] then 
      return cache[m]
    else
      local r = _fib(m - 1) + _fib(m - 2)
      cache[m] = r
      return r
    end
  end

  return _fib(n)
end

maxima = 4 * 10^6

sigma = 0
i = 0
while fib(i) < maxima do
  if fib(i) % 2 == 0 then
    sigma = sigma + fib(i)
  end
  i = i + 1
end
print(sigma)